### PR TITLE
Add a README for treetex

### PR DESCRIPTION
**Pull request by AlCutter**
### Checklist

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [X] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

### CODE REVIEW

File: docs/merkletree/treetex/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+Treetex
+=======
+
+`treetex` is a tool for drawing/visualising Merkle trees.
+
+Ensure you have latex installed on your system if you want to this; the tool
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Ensure you have `latex` command installed on your system if you want to use this; the tool
```
Also, rewrap after the modification.

File: docs/merkletree/treetex/README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+outputs `LaTeX` commands to stdout, and so needs to be used in conjuction with
+that!
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
that.
```

File: docs/merkletree/treetex/README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+Usage
+-----
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex | latex
+```
+(if you've not used that much tex - that'll create a file called `texput.pdf`)
+You can use `okular` (or your favourite PDF viewer) to view it.
+
+Use `--tree_size` to, er, set the tree size.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Use `--tree_size` flag to, er, set the tree size.
```

File: docs/merkletree/treetex/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+
+Features
+--------
+Treetex has a few tricks up its sleeveseses:
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]: Up its Sleevi?

Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Treetex has a few tricks up its sleeve:
```

File: docs/merkletree/treetex/README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+### Ephemeral nodes
+
+This tool draws trees with all the leaves at the same level as opposed to floating them up to their real level - mainly because this is how the trees in RFC6962 are drawn.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]: Wrap?
passim

File: docs/merkletree/treetex/README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+Nodes which aren't yet final are displayed with no border.
+
+### Large trees
+When drawing 'large' tree sizes, set the --megamode_threshold (default is 4) to "collapse" perfect subtrees with at least this many levels.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
When drawing 'large' tree sizes, set the `--megamode_threshold` flag (default is 4) to "collapse" perfect subtrees with at least this many levels.
```

File: docs/merkletree/treetex/README.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=100023 --megamode_threshold=4
+```
+![large tree](images/large.png)
+
+
+### Inclusion proofs
+highlight inclusion proofs
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Highlight an inclusion proof.
```

File: docs/merkletree/treetex/README.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+Choose-your-own-~~adventure~~-data.
+Takes LaTex commands too.
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --leaf_data='entry1,thing too,${\alpha}$,${\frac{bananas_{\beta}}{\delta}}$,{\LaTeX}'
+```
+
+![leaf data](images/leafdata.png)
+
+### Explain-Merkle-trees-mode
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
### Explain-Merkle-trees mode
```

File: docs/merkletree/treetex/README.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+Draw diagrams which explain what each node hash is.
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=7 --node_format=hash
+```
+
+![hash mode](images/hashmode.png)
+
+### Compact Range proofs
+Highlight node which make up (multiple) range proofs
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Highlight nodes which make up (multiple) range proofs.
```

File: docs/merkletree/treetex/README.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=18 --ranges=1:3,7:13
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]: nit (follow-up): I'd replace `7:13` range with `7:14`, for it to be less symmetric and more illustrative. Needs updating the image as well.

File: docs/merkletree/treetex/README.md (hunk 1 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: You need more than just the command installed, there's a lot of support files etc. too.

File: docs/merkletree/treetex/README.md (hunk 3 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 4 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 5 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 6 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 7 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 8 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 10 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 4 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: not done :)

**Comment by AlCutter** [comment]
PTAL

File: docs/merkletree/treetex/README.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
+Treetex
+=======
+
+`treetex` is a tool for drawing/visualising Merkle trees.
+
+Ensure you have LaTeX installed on your system if you want to this; the tool
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Ensure you have `LaTeX` installed on your system if you want to this; the tool
```

File: docs/merkletree/treetex/README.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+Features
+--------
+Treetex has a few tricks up its sleeve:
+
+### Ephemeral nodes
+
+This tool draws trees with all the leaves at the same level as opposed to
+floating them up to their real level - mainly because this is how the trees in
+RFC6962 are drawn.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
[RFC6962](https://tools.ietf.org/html/rfc6962) are drawn.
```

File: docs/merkletree/treetex/README.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+Nodes which aren't yet final are displayed with no border.
+
+### Large trees
+When drawing 'large' tree sizes, set the `--megamode_threshold` flag (default
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
When drawing large trees, set the `--megamode_threshold` flag (default
```

File: docs/merkletree/treetex/README.md [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+is 4) to "collapse" perfect subtrees with at least this many levels.
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=100023 --megamode_threshold=4
+```
+![large tree](images/large.png)
+
+
+### Inclusion proofs
+Highlight an inclusion proof.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
To highlight an inclusion proof, use the `--inclusion` flag.
```

File: docs/merkletree/treetex/README.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=23 --inclusion=13
+```
+
+![inclusion proof](images/inclusion.png)
+
+### Explicit leaf data
+Choose-your-own-~~adventure~~-data.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Choose-your-own-~~adventure~~-data mode. Use the `--leaf_data` flag to specify leaf
data explicitly.
```

File: docs/merkletree/treetex/README.md [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+Takes LaTex commands too.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]: nit: Merge this to the previous line.

File: docs/merkletree/treetex/README.md [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --leaf_data='entry1,thing too,${\alpha}$,${\frac{bananas_{\beta}}{\delta}}$,{\LaTeX}'
+```
+
+![leaf data](images/leafdata.png)
+
+### Explain-Merkle-trees mode
+Draw diagrams which explain what each node hash is.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
To draw diagrams which explain what each node hash is, use the `--node_format` flag.
```

File: docs/merkletree/treetex/README.md [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=7 --node_format=hash
+```
+
+![hash mode](images/hashmode.png)
+
+### Compact Range proofs
+Highlight node which make up (multiple) range proofs
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
To highlight nodes which make up (multiple) range proofs, use the `--ranges` flag.
```

File: docs/merkletree/treetex/README.md [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=18 --ranges=1:3,7:14
+```
+
+![compact ranges](images/compactrange.png)
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]: The node 1.6 on this picture should be highlighted.

File: docs/merkletree/treetex/README.md (hunk 11 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 12 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 15 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 18 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 17 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 19 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: TODO

File: docs/merkletree/treetex/README.md (hunk 1 above)
Comment (pavelkalinnikov) [reply docs/merkletree/treetex/README.md]: SGTM.

File: docs/merkletree/treetex/README.md (hunk 19 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: Fixed

**Comment by pavelkalinnikov** [comment]
Can't see the effect of applying suggestions. Did you commit them?

File: docs/merkletree/treetex/main.go [hunk 20]
```diff
@@ -298,28 +298,9 @@ func modifyRangeNodeInfo() error {
...
-		}
-		// Do right perfect subtree roots:
-		nBitsR := uint(bits.Len64(maskR))
-		for i, bit := nBitsR, uint64(1<<nBitsR); i > 1; i, bit = i-1, bit>>1 {
-			if maskR&bit != 0 {
-				id := compact.NewNodeID(i, p>>i)
-				modifyNodeInfo(id, func(n *nodeInfo) { n.rangeIndices = append(n.rangeIndices, ri) })
-				p += bit
-			}
+		for _, id := range compact.RangeNodes(lr[0], lr[1]) {
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/main.go]:
```suggestion
		for _, id := range compact.RangeNodes(l, r) {
```

File: docs/merkletree/treetex/main.go [hunk 21]
```diff
@@ -298,28 +298,9 @@ func modifyRangeNodeInfo() error {
...
+			modifyNodeInfo(id, func(n *nodeInfo) { n.rangeIndices = append(n.rangeIndices, ri) })
+
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/main.go]:
```suggestion
```

File: docs/merkletree/treetex/main.go (hunk 20 above)
Comment (AlCutter) [reply docs/merkletree/treetex/main.go]: done

File: docs/merkletree/treetex/README.md [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=23 --inclusion=13
+```
+
+![inclusion proof](images/inclusion.png)
+
+### Explicit leaf data
+Choose-your-own-~~adventure~~-data.
+Use the `--leaf_data` flags to specify comma-separated leaf data text explicity.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Use the `--leaf_data` flag to specify comma-separated leaf data text explicity.
```

File: docs/merkletree/treetex/README.md [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+
+```bash
+go run github.com/google/trillian/docs/merkletree/treetex --tree_size=7 --node_format=hash
+```
+
+![hash mode](images/hashmode.png)
+
+### Compact Range proofs
+To highlight nodes which make up (multiple) range proofs, use the
+`--node_format=address` flag.
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
`--ranges` flag.
```

File: docs/merkletree/treetex/README.md [hunk 24]
```diff
@@ -0,0 +1,85 @@
+Treetex
+=======
+
+`treetex` is a tool for drawing/visualising Merkle trees.
+
+Ensure you have `LaTeX` installed on your system if you want to this; the tool
```
Comment (pavelkalinnikov) [inline docs/merkletree/treetex/README.md]:
```suggestion
Ensure you have `LaTeX` installed on your system if you want to use this; the tool
```

File: docs/merkletree/treetex/README.md (hunk 23 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 22 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

File: docs/merkletree/treetex/README.md (hunk 24 above)
Comment (AlCutter) [reply docs/merkletree/treetex/README.md]: done

**Pull request merged**